emulator/clients/interrupter.c

```diff
@@ -0,0 +1,131 @@
+/**
+ * @author Magnus Mæhlum (magnmaeh@stud.ntnu.no)
+ * @brief A simple client that can be used with the emulator when the --client
+ * option is given to the emulator. It sets up two pin events that the emulator
+ * will use to set the pin high and then low again immediately after. This could
+ * be used to emulate receiving an external interrupt.
+ * 
+ * One interrupt is triggered for each character given to stdin. The user could
+ * have both the emulator and this client running in two terminals and see
+ * how the emulated FlexPRET reacts to getting interrupts.
+ * 
+ * The simplest way to transmit just one intterrupt is just to press <enter>.
+ * 
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <sys/socket.h>
+
+#include "common.h"
+#include "../../programs/lib/include/flexpret_hwconfig.h"
+
+static pin_event_t interrupt[] = {
+    { .pin = PIN_IO_INT_EXTS_0, .in_n_cycles = 10000, .high_low = HIGH },
+
+    // Wait NUM_THREADS cycles before setting low again so the HW thread gets
+    // enough time to react
+    { .pin = PIN_IO_INT_EXTS_0, .in_n_cycles = NUM_THREADS, .high_low = LOW  },
+};
+
+static void usage(int argc, char *const* argv, char *err)
+{
+    printf("Error: %s\n", err);
+    printf("Usage: %s [-a <-n <number of interrupts> -d <delay time (ms)>>]\n" \
+           "    where -a denotes 'automatic' mode\n" \
+           "          -n is the number of interrupts to send\n" \
+           "          -d is the delay between each interrupt in milliseconds\n", \
+            argv[0]
+    );
+
+    puts("\nYou wrote:\n");
+    for (int i = 0; i < argc; i++) {
+        printf("%s ", argv[i]);
+    }
+    printf("\n");
+    fflush(stdout);
+
+    exit(1);
+}
+
+int main(int argc, char *const* argv) 
+{
+    bool manual = true;
+    
+    // Only relevant when the 'automatic' mode is enabled
+    int ninterrupts = 0;
+    int delay_ms = 0;
+
+    int opt = 0;
+    
+    while((opt = getopt(argc, argv, ":an:d:")) != -1) {
+        switch (opt)
+        {
+        case 'a':
+            // 'Automatic' mode
```

Comment:
The automatic mode is very helpful for the CICD, because the client can run without any user input. The test's makefile can specify a `CLIENT_CMD`. If this is set, the command will be run together with the emulator. 

In this way, one can write a test that expects input on pins and run it with a client that sets these pins. Since it does not require user input, it can be run in CICD.